platform/security/rest/security-rest-clientapi/src/main/java/org/codice/ddf/cxf/client/ClientFactoryBuilder.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.cxf.client;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.cxf.interceptor.Interceptor;
+import org.apache.cxf.message.Message;
+import org.codice.ddf.configuration.PropertyResolver;
+
+/**
+ * Builder class for creating secure CXF client factory objects that can generate WebClients
+ *
+ * <p><b> This code is experimental. While this interface is functional and tested, it may change or
+ * be * removed in a future version of the library. </b>
+ */
+public interface ClientFactoryBuilder<T> {
+
+  SecureCxfClientFactory<T> build();
+
+  /**
+   * The endpoint the client should connect to.
+   *
+   * @param endpointUrl
+   * @return
+   */
+  ClientFactoryBuilder<T> endpointUrl(String endpointUrl);
+
+  /**
+   * The JAX-RS interface class/type that represents the endpoint
+   *
+   * @param interfaceClass
+   * @return
+   */
+  ClientFactoryBuilder<T> interfaceClass(Class<T> interfaceClass);
+
+  ClientFactoryBuilder<T> entityProviders(List<? extends Object> entityProviders);
+
+  ClientFactoryBuilder<T> interceptor(Interceptor<? extends Message> interceptor);
+
+  /**
+   * Turns the CN check off when performing a TLS/SSL connection. This is for test only.
+   *
+   * @param disableCnCheck
+   * @return
+   */
+  ClientFactoryBuilder<T> disableCnCheck(boolean disableCnCheck);
+
+  /**
+   * Whether or not to allow the client to follow redirects.
+   *
+   * @param allowRedirects
+   * @return
+   */
+  ClientFactoryBuilder<T> allowRedirects(boolean allowRedirects);
+
+  /**
+   * Time before the connection will fail with no reply.
+   *
+   * @param connectionTimeout
+   * @return
+   */
+  ClientFactoryBuilder<T> connectionTimeout(Integer connectionTimeout);
+
+  /**
+   * Time before connection will fail with no response.
+   *
+   * @param receiveTimeout
+   * @return
+   */
+  ClientFactoryBuilder<T> receiveTimeout(Integer receiveTimeout);
+
+  /**
+   * Username to use with BASIC authentication.
+   *
+   * @param username
+   * @return
+   */
+  ClientFactoryBuilder<T> username(String username);
```

Comment:
I would love to see this API combine some of these parameters such that they're not all top-level. For example:
```
builder.withBasic("username", "password").build()
```
or 
```
builder
    .withOAuth()   // Have this return some object that builds oauth configuration
        .clientId("client")
        .clientSecret("secret")
        .oauthFlow("authorization_code")
        .additionalParameters(...)
        .and()    // Return the original builder
    .build()
```

That would make it much clearer how to use the API to properly conflgure the client